src/pattern-match.js

```diff
@@ -0,0 +1,31 @@
+export function any(type) {
+  if(type === "array") {
+    return function anyMatcher(value) {
+      return Array.isArray(value);
+    };
+  } else if(type) {
+    return function anyMatcher(value) {
+      return typeof(value) === type;
+    };
+  } else {
+    return function anyMatcher() {
+      return true;
+    };
+  }
+}
+
+export function patternMatch(match) {
+  return function matcher(target) {
+    if(match === undefined) {
+      return true;
+    } else if(Array.isArray(match)) {
+      return match.every((value, index) => patternMatch(value)(target[index]));
+    } else if(typeof(match) === "object") {
+      return Object.entries(match).every(([key, value]) => patternMatch(value)(target[key]));
+    } else if(typeof(match) === "function") {
+      return match(target);
+    } else {
+      return match === target;
+    }
+  };
+}
```

Comment:
I made this a curried function, for ... no particular reason. Should it be?

Reply:
It feels right. It seems like this is simpler to "compiling" a regex and then using it repeatedly to match. Right now, all of the matching happens dynamically, and so the only overhead that we're saving on is the actual allocation of the `matcher` function, but I can imagine a world where the optimizations are significant enough that they warrant pre-allocation of some memory in order to save some computational cycles as they are matched repeatedly.

If we make the API support this right out of the gate, then there's no need to change it going back.

Fun tip: A nice way that I've discovered to make curried APIs in JavaScript is with the `=>` syntax:

```js
const patternMatch = match => target => {
  if (match === undefined) {
  }
  //...
}
```

Of course, if you need to do some work in between arguments, for the pre-allocation, then it doesn't help, but for everything else....